include/graphblas/reference/io.hpp

```diff
@@ -1076,6 +1076,80 @@ namespace grb {
 
 	} // end namespace internal::grb
 
+	template<
+		Descriptor descr = descriptors::no_operation,
+		typename DataType, typename RIT, typename CIT, typename NIT,
+		typename ValueType = DataType
+	>
+	RC set(
+		Matrix< DataType, reference, RIT, CIT, NIT  > &C,
+		const ValueType& val,
+		const Phase &phase = EXECUTE
+	) noexcept {
+#ifdef _DEBUG
+		std::cout << "Called grb::set (matrix-to-value-unmasked, reference)\n";
+#endif
+		// static checks
+		static_assert(
+			!std::is_void< DataType >::value,
+			"grb::set (unmasked set to value): cannot have a pattern "
+			"matrix as output"
+		);
+		static_assert(
+			!std::is_void< ValueType >::value,
+			"grb::set (unmasked set to value): cannot have a pattern "
+			"matrix as output"
+		);
+		NO_CAST_ASSERT( (
+			( !(descr & descriptors::no_casting)
+				&& std::is_convertible< ValueType, DataType >::value
+			) || std::is_same< DataType, ValueType >::value
+			), "grb::set (unmasked set to value): ",
+			"called with non-matching value types"
+		);
+
+		// dynamic checks
+		assert( phase != TRY );
+
+		// delegate
+		if( phase == RESIZE ) {
+			return SUCCESS;
+		} else {
+			assert( phase == EXECUTE );
+
+			const size_t m = nrows( C );
+			const size_t n = ncols( C );
+			const size_t nz = nnz( C );
+
+			if( m == 0 || n == 0 || nz == 0 ) { return SUCCESS; }
+
+#ifdef _H_GRB_REFERENCE_OMP_IO
+			#pragma omp parallel default( none ) \
+				shared( C ) firstprivate( val, m, n, nz )
+#endif
+			{
+				{ // CRS section
+					size_t range = internal::getCRS( C ).copyFromRange( nz, m );
+					size_t start = 0, end = range;
+#ifdef _H_GRB_REFERENCE_OMP_IO
+					config::OMP::localRange( start, end, 0, range );
+#endif
+					internal::getCRS( C ).assignValue( nz, m, start, end, val );
```

Comment:
I think this sets all current values to the given one, rather than all values in the matrix. This would be the correct implementation for masked sets, in particular, for `grb::set( C, C, val )`, where the second `C` indicates the matrix itself acts as the mask for the set operation. For `grb::set( C, val )` I would expect the entire matrix `C` to have the given `val` everywhere. (Also for the CCS section ofc.)

Btw if implementing the masked variants, take care to also consider the descriptors that modify the mask behaviours, which could get tricky

Reply:
The idea of this primitive (at leat in my head) was to assign the *existing* non-zero values to the given one. Which would be an alias for `grb::set( input/output=A, mask=A, value=x)`